### PR TITLE
Include short sha in log messages and error urls

### PR DESCRIPTION
**Pull request by yaahc**
## Motivation

As we approach our alpha release we've decided we want to plan ahead for the user bug reports we will eventually receive. One of the bigger issues we foresee is determining exactly what version of the software users are running, and particularly how easy it may or may not be for users to accidentally discard this information when reporting bugs.

## Solution

To defend against this, we've decided to include the exact git sha for any given build in the compiled artifact. This information will then be re-exported as a span early in the application startup process, so that all logs and error messages should include the sha as their very first span. We've also added this sha as issue metadata for `color-eyre`'s github issue url auto generation feature, which should make sure that the sha is easily available in bug reports we receive, even in the absence of logs.

The code in this pull request has:
  - [ ] Documentation Comments
  - [ ] Unit Tests and Property Tests

## Related Issues

Partly implements https://github.com/ZcashFoundation/zebra/issues/692

### CODE REVIEW

**Comment by hdevalence** [comment]
What does the output look like?

Should we consider changing the span name from `zebrad` to something shorter, to minimize the extra data attached to every line?

Also, should we set the span level above info, so that the sha shows up on errors as well?

**Comment by yaahc** [comment]
> What does the output look like?

```
Nov 30 16:47:28.745  INFO zebrad{GIT_SHA=7e6c588a}:sync: zebrad::components::sync: exhausted prospective tip set
```

> Should we consider changing the span name from zebrad to something shorter, to minimize the extra data attached to every line?

/shrug

> Also, should we set the span level above info, so that the sha shows up on errors as well?

`SpanTrace` follows the exact same span filtering rules that our events do, so this will show up in error reports if the filter level is set to info or above, but I do think switching the level to Error would be a good idea so this doesn't get filtered out when someone sets quieter log levels.

**Comment by yaahc** [comment]
tweeked it, now it looks like this

```
Nov 30 17:07:37.177  INFO git{SHA=7e6c588a}:peer{addr=47.90.200.114:8233}:handle_message_as_request{msg=GetHeaders { known_blocks: zebra_chain::block::hash::Hash, len=32, stop: None }}:inbound:state: zebra_state::service: responding to peer GetBlocks or GetHeaders final_height=Height(12312) response_len=160 chain_tip_height=Height(103089) stop_height=None intersection_height=Some(Height(12152))
```

**Comment by hdevalence** [comment]
What if we set the span name to `""` and made the key `zebrad`, so that it looked like `{zebrad=7e6c588a}`? Then if we put a tag in there or something it would become `{zebrad=1.3.8}` or whatever.

EDIT: didn't realize I was racing your comment, the output you posted looks good, this is just a bikeshed

File: zebrad/build.rs [hunk 1]
```diff
@@ -0,0 +1,12 @@
+extern crate vergen;
+
+use vergen::{generate_cargo_keys, ConstantsFlags};
+
+fn main() {
+    // Setup the flags, toggling off the 'SEMVER_FROM_CARGO_PKG' flag
+    let mut flags = ConstantsFlags::empty();
+    flags.toggle(ConstantsFlags::SHA_SHORT);
+
```
Comment (teor2345) [inline zebrad/build.rs:10]: Make sure that every build gets the latest commit hash
```suggestion
    flags.toggle(ConstantsFlags::REBUILD_ON_HEAD_CHANGE);

```

File: zebrad/src/application.rs [hunk 2]
```diff
@@ -92,6 +92,7 @@ impl Application for ZebradApp {
         color_eyre::config::HookBuilder::default()
             .issue_url(concat!(env!("CARGO_PKG_REPOSITORY"), "/issues/new"))
             .add_issue_metadata("version", env!("CARGO_PKG_VERSION"))
+            .add_issue_metadata("git sha", ZebradCmd::SHA)
```
Comment (teor2345) [inline zebrad/src/application.rs]: Git uses `SHA1`s for a bunch of different object types. This is the `SHA1` of the git `HEAD` commit:
```suggestion
            .add_issue_metadata("git commit", ZebradCmd::GIT_COMMIT)
```

File: zebrad/src/commands.rs [hunk 3]
```diff
@@ -38,6 +38,8 @@ pub enum ZebradCmd {
 }
 
 impl ZebradCmd {
+    pub const SHA: &'static str = env!("VERGEN_SHA_SHORT");
```
Comment (teor2345) [inline zebrad/src/commands.rs]: Git uses `SHA1`s for a bunch of different object types. This is the `SHA1` of the git `HEAD` commit:
```suggestion
    pub const GIT_COMMIT: &'static str = env!("VERGEN_SHA_SHORT");
```

File: zebrad/src/commands.rs [hunk 4]
```diff
@@ -50,6 +52,19 @@ impl ZebradCmd {
     }
 }
 
+impl Runnable for ZebradCmd {
+    fn run(&self) {
+        let span = error_span!("git", SHA = %Self::SHA);
```
Comment (teor2345) [inline zebrad/src/commands.rs]: Git uses `SHA1`s for a bunch of different object types. This is the `SHA1` of the git `HEAD` commit:
```suggestion
        let span = error_span!("git", commit = %Self::GIT_COMMIT);
```

**Comment by teor2345** [comment]
I removed the `closes #692`, because there's still a bunch of work listed in #692. We should make an explicit decision about when we want to do this work, rather than accidentally ignoring it.

**Comment by hdevalence** [comment]
I think the span should be of the form `{zebrad=short_sha}` (i.e., empty span name, field named `zebrad`). This is concise and identifies that we are running `zebrad` (not, e.g., git) and is compatible with tagged versions in the future.

**Review by hdevalence — DISMISSED** [review]
LGTM

**Review by hdevalence — APPROVED** [review]
LGTM2

**Comment by dconnolly** [comment]
I have realized this doesn't work inside Cloud Build because they tarball up your source and inject the git information via environmental variables, but you are not in a git repo:

image
image
image

**Comment by yaahc** [comment]
where does that `Built-in substitutions` description come from?

**Comment by dconnolly** [comment]
> where does that `Built-in substitutions` description come from?

The app hook pre-PR merge and  `gcloud build submit` post-PR merge

**Comment by dconnolly** [comment]
In both, we have $SHORT_SHA precomputed and available:

image


We should probably inspect env for that first and if it's not found use vergen

File: zebrad/src/application.rs [hunk 5]
```diff
@@ -43,6 +43,10 @@ pub struct ZebradApp {
     state: application::State<Self>,
 }
 
+impl ZebradApp {
+    pub const GIT_COMMIT: &'static str = env!("VERGEN_SHA_SHORT");
```
Comment (dconnolly) [inline zebrad/src/application.rs:47]:
```
env!("SHORT_SHA") || env!("VERGEN_SHA_SHORT");
```

or something like that

Comment (yaahc) [reply zebrad/src/application.rs:47]: fixed in https://github.com/ZcashFoundation/zebra/pull/1442/files hopefully

**Comment by teor2345** [comment]
> I have realized this doesn't work inside Cloud Build because they tarball up your source and inject the git information via environmental variables, but you are not in a git repo:

I updated #692 with this info, and a link to the Google Cloud reference docs: https://cloud.google.com/cloud-build/docs/configuring-builds/substitute-variable-values#using_default_substitutions

**Pull request merged**